contracts/src/utils/math/alloy.rs

```diff
@@ -0,0 +1,167 @@
+        // We can refine our estimation by noticing that the middle of that
+        // interval minimizes the error. If we move x_n to equal
+        // 2**(e-1) + 2**(e-2), then we reduce the error to ε_n ≤
+        // 2**(e-2). This is going to be our x_0 (and ε_0)
+        xn = (uint!(3_U256) * xn) >> 1; // ε_0 := | x_0 - sqrt(a) | ≤ 2**(e-2)
+
+        // From here, Newton's method give us:
+        // x_{n+1} = (x_n + a / x_n) / 2
+        //
+        // One should note that:
+        // x_{n+1}² - a = ((x_n + a / x_n) / 2)² - a
+        //              = ((x_n² + a) / (2 * x_n))² - a
+        //              = (x_n⁴ + 2 * a * x_n² + a²) / (4 * x_n²) - a
+        //              = (x_n⁴ + 2 * a * x_n² + a² - 4 * a * x_n²) / (4 * x_n²)
+        //              = (x_n⁴ - 2 * a * x_n² + a²) / (4 * x_n²)
+        //              = (x_n² - a)² / (2 * x_n)²
+        //              = ((x_n² - a) / (2 * x_n))²
+        //              ≥ 0
+        // Which proves that for all n ≥ 1, sqrt(a) ≤ x_n
+        //
+        // This gives us the proof of quadratic convergence of the sequence:
+        // ε_{n+1} = | x_{n+1} - sqrt(a) |
+        //         = | (x_n + a / x_n) / 2 - sqrt(a) |
+        //         = | (x_n² + a - 2*x_n*sqrt(a)) / (2 * x_n) |
+        //         = | (x_n - sqrt(a))² / (2 * x_n) |
+        //         = | ε_n² / (2 * x_n) |
+        //         = ε_n² / | (2 * x_n) |
+        //
+        // For the first iteration, we have a special case where x_0 is known:
+        // ε_1 = ε_0² / | (2 * x_0) |
+        //     ≤ (2**(e-2))² / (2 * (2**(e-1) + 2**(e-2)))
+        //     ≤ 2**(2*e-4) / (3 * 2**(e-1))
+        //     ≤ 2**(e-3) / 3
+        //     ≤ 2**(e-3-log2(3))
+        //     ≤ 2**(e-4.5)
+        //
+        // For the following iterations, we use the fact that, 2**(e-1) ≤
+        // sqrt(a) ≤ x_n: ε_{n+1} = ε_n² / | (2 * x_n) |
+        //         ≤ (2**(e-k))² / (2 * 2**(e-1))
+        //         ≤ 2**(2*e-2*k) / 2**e
+        //         ≤ 2**(e-2*k)
+        xn = (xn + a / xn) >> 1; // ε_1 := | x_1 - sqrt(a) | ≤ 2**(e-4.5)  -- special case, see above
+        xn = (xn + a / xn) >> 1; // ε_2 := | x_2 - sqrt(a) | ≤ 2**(e-9)    -- general case with k = 4.5
+        xn = (xn + a / xn) >> 1; // ε_3 := | x_3 - sqrt(a) | ≤ 2**(e-18)   -- general case with k = 9
+        xn = (xn + a / xn) >> 1; // ε_4 := | x_4 - sqrt(a) | ≤ 2**(e-36)   -- general case with k = 18
+        xn = (xn + a / xn) >> 1; // ε_5 := | x_5 - sqrt(a) | ≤ 2**(e-72)   -- general case with k = 36
+        xn = (xn + a / xn) >> 1; // ε_6 := | x_6 - sqrt(a) | ≤ 2**(e-144)  -- general case with k = 72
+
+        // Because e ≤ 128 (as discussed during the first estimation phase), we
+        // know have reached a precision ε_6 ≤ 2**(e-144) < 1. Given
+        // we're operating on integers, then we can ensure that xn is
+        // now either sqrt(a) or sqrt(a) + 1.
+        xn - U256::from(xn > a / xn)
+    }
+
+    fn average(self, rhs: Self) -> Self {
```

Comment:
Average implemented this way to avoid overflow.
[Solidity example](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/442886ed5ff8a0b9ab477b191f5238541ee6d772/contracts/utils/math/Math.sol#L110).

Reply:
smart